### PR TITLE
[mono] Runtime cleanup functions removal

### PR DESCRIPTION
**Pull request by CoffeeFlux**
Is the second commit worth it? The first one definitely is.

It is probably worth checking through everything I'm removing before approving, in case any of it might be useful down the line. I don't think so, but a second pair of eyes would be good. 

There's no nice way to split that second commit up, sorry. This is just a huge diff.

### CODE REVIEW

File: src/mono/mono/metadata/profiler-events.h [hunk 1]
```diff
@@ -32,8 +32,6 @@
 #endif
 
 MONO_PROFILER_EVENT_0(runtime_initialized, RuntimeInitialized)
-MONO_PROFILER_EVENT_0(runtime_shutdown_begin, RuntimeShutdownBegin)
```
Comment (vargaz) [inline src/mono/mono/metadata/profiler-events.h:35]: Isn't this part of the embedding API ?

**Review by vargaz — APPROVED** [review]

File: src/mono/mono/metadata/profiler-events.h (hunk 1 above)
Comment (CoffeeFlux) [reply src/mono/mono/metadata/profiler-events.h:35]: I thought that one isn't public, but if it is maybe I need to undo that; will check.

Comment (lambdageek) [reply src/mono/mono/metadata/profiler-events.h:35]: I think profiler events have API guarantees

**Review by lambdageek — COMMENTED** [review]
I'm not sure removing the cleanup functionality is a good idea.
There are embedders (wine I think is one) that use us in plugins, so they shutdown and start the runtime multiple times.  I think we could move more code under `DISABLE_CLEANUP`, but I'm not sure we should lose the ability to cleanup entirely.  (And perhaps we should leave it compiled in on desktop).

**Comment by CoffeeFlux** [comment]
The main argument for getting rid of it is that it doesn't work today (and I don't think it's even close, we'll leak a whole lot of memory) and isn't a configuration we're interested in supporting. I can move this code under the ifdef instead if you think the functionality might be useful down the line, but I was under the impression that it probably is not.

**Comment by vargaz** [comment]
It is not actually possible to shut down and restart the runtime, we don't free/unload everything, static variables are not reinitialized, signal handlers not unregistered, etc.. Maybe by putting the runtime in a shared lib, and unloading/reloading it would work.

**Comment by lambdageek** [comment]
ok, well if it doesn't work correctly and unloading the shared library is enough, maybe that's alright.

**Pull request merged**